# **Assignment 2**

1. PANAGIOTIS KAFANTARIS **p3352328**
2. IOANNIS MICHALOPOULOS **p3352314**

**Description:**

**Step 1:** Join one of the empty teams on e-class. Up to 2 people in each team.


**Step 2:** Get access to the Instagram Influencer Dataset.

**Step 3:** Get  a random sample of 20,000 Instagram posts from the dataset. Save it in a csv file that includes the info for 1 post in each row.

**Step 4:** Create a Colab Notebook that:

* Reads the sampled dataset from Step 2.
* Uses the raw sampled dataset in the sampled dataset to engineer features.
* Splits into training, testing, and validation.
* Uses the training set to train a model that can predict the number of likes of a given post.
* Uses the validation set for parameter tuning
* Uses the testing set to evaluate the final model's predictive power.

We download the libraries

In [ ]:
!pip install langdetect
!pip install googletrans
!pip install spacy
!python -m spacy download en_core_web_sm

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 981.5/981.5 kB 12.3 MB/s eta 0:00:00
  Preparing metadata (setup.py) ... done
  Created wheel for langdetect: filename=langdetect-1.0.9-py3-none-any.whl size=993222 sha256=81fae2cf75a60e31ecf4e80cae485a28a1854672f061360c2e148faa8d5dade1
  Stored in directory: /root/.cache/pip/wheels/0a/f2/b2/e5ca405801e05eb7c8ed5b3b4bcf1fcabcd6272c167640072e
Successfully built langdetect
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 57.5/57.5 kB 2.2 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 12.8/12.8 MB 43.1 MB/s eta 0:00:00
✔ Download and installation successful
You can now load the package via spacy.load('en_core_web_sm')
⚠ Restart to reload dependencies
If you are in a Jupyter or Colab notebook, you may need to restart Python in
order to load all the package's dependencies. You can do this by selecting the
'Restart kernel' or 'Restart runtime' option.


# **INTRODUCTION:**

In this assignment, we utilize the [instagram influencer dataset](https://sites.google.com/site/sbkimcv/dataset/instagram-influencer-dataset) containing millions of posts. Our goal is to build a predictive model that estimates the number of likes a post might receive based on its metadata.

To achieve this, we started by extracting a manageable sample of posts from the dataset, which was stored in a compressed format. The sample was unzipped and manually uploaded to Google Drive for ease of manipulation and analysis.

From the extracted data, we constructed a clean and structured dataset by processing the metadata associated with each post. This includes information such as captions, user details, location, and engagement metrics. The processed data was saved as CSV files to ensure compatibility with various tools and workflows. These datasets serve as the foundation for training and evaluating our predictive model.

## **CONNECTION TO GOOGLE DRIVE**

Everything was placed under google drive storage and as a result we must connect and navigate to the Instagram influencer dataset folder

In [ ]:
import pandas as pd
from google.colab import drive

In [ ]:
drive.mount('/content/drive')

Mounted at /content/drive


In [ ]:
!ls '/content/drive/My Drive/Colab Notebooks/Instagram influencer dataset'

influencers.txt		      Post_images    README.docx
JSON-Image_files_mapping.txt  Post_metadata  sample_images.zip


# **LOAD INFLUENCERS**

In [ ]:
file_path = '/content/drive/My Drive/Colab Notebooks/Instagram influencer dataset/influencers.txt'

df_influencers = pd.read_csv(file_path, sep="\t", skiprows=[1])

In [ ]:
df_influencers.shape

(33935, 5)

In [ ]:
output_csv_path = "influencers_data.csv"

In [ ]:
df_influencers.to_csv(output_csv_path, index=False)

print(f"Data successfully written to {output_csv_path}")

Data successfully written to influencers_data.csv


In [ ]:
df_influencers = pd.read_csv(output_csv_path)

In [ ]:
df_influencers.head()

Username Category  #Followers  #Followees  #Posts
0        makeupbynvs   beauty        1432        1089     363
1  jaquelinevandoski   beauty      137600         548     569
2      anisaartistry   beauty       64644         289     391
3  rubina_muartistry   beauty      496406         742     887
4        beautyxabbi   beauty        2050        1423     751

# **Extracting a Sample of Posts**

The following code is designed to extract a sample of posts from the Instagram influencer dataset. The process involves reading and filtering through a large volume of data to isolate a subset that meets our requirements. Once the desired number of posts was identified, we manually interrupted the process to ensure efficiency and accuracy.

The extracted sample was then saved into a folder named extracted_20k_records within Google Drive for subsequent analysis and manipulation. This step was critical in managing the dataset size and focusing our efforts on a representative subset.

**Note:** This code is computationally intensive and time-consuming; therefore, it should not be executed unless necessary.

In [ ]:
!apt-get install -y p7zip-full

Reading package lists... Done
Building dependency tree... Done
Reading state information... Done
p7zip-full is already the newest version (16.02+dfsg-8).
0 upgraded, 0 newly installed, 0 to remove and 49 not upgraded.


## **DONT RUN **

In [ ]:
!mkdir -p /content/temp_extracted/

# Extract the archive using 7z into the temporary directory
!7z x '/content/drive/My Drive/Colab Notebooks/Instagram influencer dataset/Post_metadata/posts_info.zip' -o/content/temp_extracted/

In [ ]:
##  List the files in the extracted directory
!ls /content/temp_extracted/

# Check for your specific file
!ls /content/temp_extracted/info/

In [ ]:
!cp -r /content/temp_extracted/ '/content/drive/My Drive/Colab Notebooks/extracted_20k_records/'

An example of how our data looks like

In [ ]:
import json

file_path = '/content/drive/My Drive/Colab Notebooks/extracted_20k_records/info/beckilw_sw-1684125244807239621.info'

# Read and parse the file contents
with open(file_path, 'r') as file:
    data = file.read()

# Parse the string into a dictionary
try:
    json_data = json.loads(data)
    for key, value in json_data.items():
        print(f"{key}: {value}")
except json.JSONDecodeError as e:
    print("Error decoding JSON:", e)

gating_info: None
viewer_can_reshare: True
display_resources: [{'src': 'https://scontent-lax3-1.cdninstagram.com/vp/a4e28f66221af03382469542660d9617/5DA9611C/t51.2885-15/sh0.08/e35/s640x640/26156580_544005902617468_5029808145619222528_n.jpg?_nc_ht=scontent-lax3-1.cdninstagram.com', 'config_width': 640, 'config_height': 640}, {'src': 'https://scontent-lax3-1.cdninstagram.com/vp/1f49ea763af304914e0ab6a85f7a7c07/5DA3081C/t51.2885-15/sh0.08/e35/s750x750/26156580_544005902617468_5029808145619222528_n.jpg?_nc_ht=scontent-lax3-1.cdninstagram.com', 'config_width': 750, 'config_height': 750}, {'src': 'https://scontent-lax3-1.cdninstagram.com/vp/0483ef0a397a594cd0115f19ca1e2b55/5DBF81F9/t51.2885-15/e35/26156580_544005902617468_5029808145619222528_n.jpg?_nc_ht=scontent-lax3-1.cdninstagram.com', 'config_width': 1080, 'config_height': 1080}]
viewer_in_photo_of_you: False
viewer_has_saved_to_collection: False
viewer_has_saved: False
owner: {'username': 'beckilw_sw', 'is_unpublished': False, 'request

## **POSTS TO DATAFRAME**

In this section, we process Instagram posts stored in JSON format to transform them into a structured DataFrame. Each post's features are extracted and organized into corresponding columns, ensuring the data is both accessible and manageable for analysis.

The below code carefully handles nested structures and potential missing values, assigning default values where necessary. This transformation allows us to work with a clean and consistent dataset, simplifying further processing and enabling effective model training.

In [ ]:
import os
import json
import re

In [ ]:
dir_path = '/content/drive/My Drive/Colab Notebooks/extracted_20k_records/info/'

In [ ]:
post_data = []

# Iterate through all .info files in the directory
for filename in os.listdir(dir_path):
  if filename.endswith(".info"):
    file_path = os.path.join(dir_path, filename)

    # Open and read the .info file
    with open(file_path, 'r') as file:
      try:
        # Parse JSON content
        post = json.load(file)

        # Extract relevant fields
        post_data.append({
          'username': post.get('owner', {}).get('username', ''),  # Default to empty string if missing
          'post_id': post.get('id', ''),  # Default to empty string if missing
          'caption': (post.get('edge_media_to_caption', {})
                    .get('edges', [{}])[0]
                    .get('node', {})
                    .get('text', '') if post.get('edge_media_to_caption', {}).get('edges') else ''), # Handle deeply nested structure with defaults
          'likes': post.get('edge_media_preview_like', {}).get('count', 0),  # Default to 0 if missing
          'comments': post.get('edge_media_preview_comment', {}).get('count', 0),  # Default to 0 if missing
          'comments_parent': post.get('edge_media_to_parent_comment', {}).get('count', 0),  # Default to 0
          'timestamp': post.get('taken_at_timestamp', None),  # Default to None if missing
          'sponsorship': bool(post.get('edge_media_to_sponsor_user', {}).get('edges', [])),  # Boolean conversion
          'usertags': len(post.get('edge_media_to_tagged_user', {}).get('edges', [])),  # Count tagged users
          'location': post.get('location', ''),
          'accessibility_caption': post.get('accessibility_caption', ''),  # Default to empty string
          'is_verified': post.get('owner', {}).get('is_verified', False),  # Default to False
        })

        location = post.get('location')  # This will be None if location doesn't exist
        if location and isinstance(location.get('address_json'), str):
            try:
                location_data = json.loads(location.get('address_json', ''))
                post_data[-1]['location'] = location_data.get('country_code', '')
            except json.JSONDecodeError:
                post_data[-1]['location'] = ''  # Fallback if JSON is invalid
        else:
            post_data[-1]['location'] = ''  # If location is None or address_json is not a valid string

      except json.JSONDecodeError:
          print(f"Error decoding JSON in file: {filename}")
      except Exception as e:
          print(f"Unexpected error processing file {filename}: {e}")


# Create a DataFrame from the list of posts
df_posts = pd.DataFrame(post_data)

In [ ]:
df_posts.head()

username              post_id  \
0  susannahkellogg  1992977640461680979   
1       la.brandon  1673289887866232073   
2        kailovera  1071828292149529960   
3      cowboytails  1735717259248626469   
4            noore  1732054186861771249   

                                             caption  likes  comments  \
0  Some days I just live for naptime!  How many o...    165         0   
1  Winter essential. I use this every other day t...    345         7   
2  Saliendo... Vaya fecha para tomar un vuelo... ...     59         0   
3  From last night’s birthday festivities @luckyd...    112         5   
4         Hey I’m at the store do you need anything?   2194       299   

   comments_parent   timestamp  sponsorship  usertags location  \
0                0  1551801470        False         1       US   
1                7  1513691720        False         1            
2                0  1441991910        False         0       VE   
3                5  1521133643        False         0       US   
4              299  1520696970        False         1       US   

                   accessibility_caption  is_verified  
0  Image may contain: one or more people        False  
1        No photo description available.        False  
2        No photo description available.        False  
3      Image may contain: dog and indoor        False  
4     Image may contain: 1 person, shoes        False

In [ ]:
len(df_posts)

20323

Write dataset to CSV

In [ ]:
output_csv_path = "posts_data.csv"

In [ ]:
df_posts.to_csv(output_csv_path, index=False)

print(f"Data successfully written to {output_csv_path}")

Data successfully written to posts_data.csv


Read to DataFrame

In [ ]:
df_posts = pd.read_csv(output_csv_path)

Megre influencers dataset with Posts

In [ ]:
df_merged_posts_influencers = pd.merge(
    df_posts,
    df_influencers,
    left_on='username',
    right_on='Username',
    how='inner'
)

In [ ]:
len(df_merged_posts_influencers)

19997

In [ ]:
output_csv_path = "merge_data.csv"

In [ ]:
df_merged_posts_influencers.to_csv(output_csv_path, index=False)

print(f"Data successfully written to {output_csv_path}")

Data successfully written to merge_data.csv


# **Feature Engineering for Model Readiness**

In this section, we process our data to extract and construct features that will enhance the model's ability to comprehend the dataset. Since many columns in the dataset contain strings or boolean values, we need to transform these into numerical representations or new derived features that capture meaningful information.

By creating additional columns with structured and model-friendly data, we enable the model to learn from patterns that might otherwise be obscured in raw formats. This step is crucial for improving predictive accuracy and ensuring the dataset is fully prepared for machine learning tasks.

In [ ]:
import pandas as pd
from datetime import datetime

data_path = "merge_data.csv"
df_merged = pd.read_csv(data_path)

This code transforms raw timestamps into meaningful categories by decoding Unix timestamps into readable dates and assigning them to time-of-day buckets **(Morning, Afternoon, Evening, Night)** and seasons **(Winter, Spring, Summer, Autumn)** based on the hour and month. These categorizations help analyze temporal patterns and trends in user activity or post engagement, and will be used as features for the predictive model to improve its accuracy and predictions.

In [ ]:
df_merged['decoded_timestamp'] = df_merged['timestamp'].apply(lambda x: datetime.utcfromtimestamp(x).strftime('%Y-%m-%d %H:%M:%S') if pd.notnull(x) else None)

# Define a function to assign time-of-day buckets
def assign_time_bucket(decoded_timestamp):
    if pd.isnull(decoded_timestamp):
        return None
    hour = int(datetime.strptime(decoded_timestamp, '%Y-%m-%d %H:%M:%S').hour)
    if 5 <= hour < 12:
        return 'Morning'
    elif 12 <= hour < 17:
        return 'Afternoon'
    elif 17 <= hour < 21:
        return 'Evening'
    else:
        return 'Night'

# Define a function to assign a season based on the month
def assign_season(decoded_timestamp):
    if pd.isnull(decoded_timestamp):
        return None
    month = int(datetime.strptime(decoded_timestamp, '%Y-%m-%d %H:%M:%S').month)
    if month in [12, 1, 2]:
        return 'Winter'
    elif month in [3, 4, 5]:
        return 'Spring'
    elif month in [6, 7, 8]:
        return 'Summer'
    else:
        return 'Autumn'


# Create a new column for time-of-day buckets
df_merged['time_of_day'] = df_merged['decoded_timestamp'].apply(assign_time_bucket)
# Create a new column for seasons
df_merged['season'] = df_merged['decoded_timestamp'].apply(assign_season)

One Hot Encoding for the categorical columns

In [ ]:
df_merged = pd.get_dummies(df_merged, columns=["Category"], prefix="Category")
df_merged = pd.get_dummies(df_merged, columns=["time_of_day"], prefix="time_of_day")
df_merged = pd.get_dummies(df_merged, columns=["season"], prefix="season")

Turn boolean variables into numerics

In [ ]:
df_merged = df_merged.astype({col: 'int' for col in df_merged.select_dtypes(include='bool').columns})

In [ ]:
columns_to_convert = ['sponsorship', 'is_verified']
for col in columns_to_convert:
    df_merged[col] = df_merged[col].astype(bool).astype(int)

This feature introduces a location-based trend into the dataset, allowing the model to learn whether certain locations have a higher likelihood of receiving more likes.

In [ ]:
location_likes_mean = df_merged.groupby('location')['likes'].mean().to_dict()
df_merged['location_likes_mean'] = df_merged['location'].map(location_likes_mean)

In [ ]:
from textblob import TextBlob

In [ ]:
df_merged['caption'] = df_merged['caption'].fillna('')

To enrich our dataset with meaningful insights from the caption column, we derive the following features:

1. caption_length: The total number of characters in the caption.
2. word_count: The number of words in the caption.
3. hashtag_count: The count of hashtags (#) in the caption.
4. mention_count: The count of mentions (@) in the caption.

In [ ]:
df_merged['caption_length'] = df_merged['caption'].apply(len)
df_merged['word_count'] = df_merged['caption'].apply(lambda x: len(x.split()))
df_merged['hashtag_count'] = df_merged['caption'].apply(lambda x: x.count('#'))
df_merged['mention_count'] = df_merged['caption'].apply(lambda x: x.count('@'))

In [ ]:
df_merged.head()

username              post_id  \
0  susannahkellogg  1992977640461680979   
1       la.brandon  1673289887866232073   
2        kailovera  1071828292149529960   
3      cowboytails  1735717259248626469   
4            noore  1732054186861771249   

                                             caption  likes  comments  \
0  Some days I just live for naptime!  How many o...    165         0   
1  Winter essential. I use this every other day t...    345         7   
2  Saliendo... Vaya fecha para tomar un vuelo... ...     59         0   
3  From last night’s birthday festivities @luckyd...    112         5   
4         Hey I’m at the store do you need anything?   2194       299   

   comments_parent   timestamp  sponsorship  usertags location  ...  \
0                0  1551801470            0         1       US  ...   
1                7  1513691720            0         1      NaN  ...   
2                0  1441991910            0         0       VE  ...   
3                5  1521133643            0         0       US  ...   
4              299  1520696970            0         1       US  ...   

  time_of_day_Night  season_Autumn season_Spring  season_Summer  \
0                 0              0             1              0   
1                 0              0             0              0   
2                 0              1             0              0   
3                 0              0             1              0   
4                 0              0             1              0   

   season_Winter  location_likes_mean caption_length  word_count  \
0              0          3370.577093            457          86   
1              1                  NaN            170          29   
2              0            62.333333             96          13   
3              0          3370.577093            342          38   
4              0          3370.577093             42           9   

   hashtag_count  mention_count  
0              2              1  
1              1              1  
2              6              0  
3             31              1  
4              0              0  

[5 rows x 39 columns]

In [ ]:
from langdetect import detect
from googletrans import Translator

To ensure that the captions are uniformly processed, we detect their language and translate non-English captions to English. The following steps are performed:

1. Language Detection: The detect function identifies the language of the caption.
2. Translation: If the caption is not in English, it is translated to English using the Translator class.
3. Fallback Mechanism: If language detection or translation fails, the original caption is retained.

The transformed captions are then stored back in the caption column, ensuring that the dataset is ready for further processing and model training. This step helps the model by standardizing the textual input, improving its ability to learn from the captions.

In [ ]:
translator = Translator()

def detect_and_translate(caption):
    try:
        lang = detect(caption)
        if lang != 'en':
            caption = translator.translate(caption, src=lang, dest='en').text
        return caption
    except:
        return caption  # Return original if detection or translation fails

df_merged['caption'] = df_merged['caption'].fillna('').apply(detect_and_translate)

<ipython-input-64-826c3cac82cc>:10: RuntimeWarning: coroutine 'Translator.translate' was never awaited
  return caption  # Return original if detection or translation fails


In [ ]:
import re
from nltk.corpus import stopwords
import string

import nltk

Cleaning Captions

In [ ]:
nltk.download('stopwords')

stop_words = set(stopwords.words('english'))

def clean_caption(caption):
    # Remove URLs
    caption = re.sub(r'http\S+|www\S+|https\S+', '', caption, flags=re.MULTILINE)
    # Remove punctuation
    caption = caption.translate(str.maketrans('', '', string.punctuation))
    # Remove emojis
    caption = re.sub(r'[^\x00-\x7F]+', '', caption)
    # Remove stopwords
    caption = ' '.join([word for word in caption.split() if word.lower() not in stop_words])
    return caption

df_merged['caption_cleaned'] = df_merged['caption'].apply(clean_caption)


[nltk_data] Downloading package stopwords to /root/nltk_data...
[nltk_data]   Package stopwords is already up-to-date!


### **Extracting Sentiment Features**

Sentiment features provide the model with an understanding of the emotional tone and personal nature of the captions. This information can influence the number of likes a post receives, as positive or engaging captions are more likely to resonate with an audience.

In [ ]:
def extract_features(caption):
    features = {}
    sentiment = TextBlob(caption).sentiment
    features['sentiment_polarity'] = sentiment.polarity  # -1 (negative) to 1 (positive)
    features['sentiment_subjectivity'] = sentiment.subjectivity  # 0 (objective) to 1 (subjective)
    return features

caption_features = df_merged['caption_cleaned'].apply(extract_features).apply(pd.Series)
df_merged = pd.concat([df_merged, caption_features], axis=1)


In [ ]:
df_merged.head()

username              post_id  \
0  susannahkellogg  1992977640461680979   
1       la.brandon  1673289887866232073   
2        kailovera  1071828292149529960   
3      cowboytails  1735717259248626469   
4            noore  1732054186861771249   

                                             caption  likes  comments  \
0  Some days I just live for naptime!  How many o...    165         0   
1  Winter essential. I use this every other day t...    345         7   
2  Saliendo... Vaya fecha para tomar un vuelo... ...     59         0   
3  From last night’s birthday festivities @luckyd...    112         5   
4         Hey I’m at the store do you need anything?   2194       299   

   comments_parent   timestamp  sponsorship  usertags location  ...  \
0                0  1551801470            0         1       US  ...   
1                7  1513691720            0         1      NaN  ...   
2                0  1441991910            0         0       VE  ...   
3                5  1521133643            0         0       US  ...   
4              299  1520696970            0         1       US  ...   

  season_Summer  season_Winter location_likes_mean  caption_length  \
0             0              0         3370.577093             457   
1             0              1                 NaN             170   
2             0              0           62.333333              96   
3             0              0         3370.577093             342   
4             0              0         3370.577093              42   

   word_count  hashtag_count mention_count  \
0          86              2             1   
1          29              1             1   
2          13              6             0   
3          38             31             1   
4           9              0             0   

                                     caption_cleaned  sentiment_polarity  \
0  days live naptime many mamas relate ad Caleb r...             0.32197   
1  Winter essential use every day help keep face ...             0.20000   
2  Saliendo Vaya fecha para tomar un vuelo trip l...             0.00000   
3  last nights birthday festivities luckydogbar d...             0.40000   
4                         Hey Im store need anything             0.00000   

   sentiment_subjectivity  
0                0.520000  
1                0.400000  
2                0.000000  
3                0.533333  
4                0.000000  

[5 rows x 42 columns]

The process involves cleaning and standardizing Instagram captions by removing unnecessary text (e.g., prefixes, placeholders, numbers) and filtering out common stopwords. **Lemmatization** is applied using `SpaCy` to reduce words to their base forms (e.g., "people" to "person") and eliminate redundant variations. This results in a simplified, meaningful  `cleaned_caption` column, ready for analysis or modeling with reduced noise and improved focus on relevant content.

In [ ]:
from nltk.tokenize import word_tokenize
import spacy

nltk.download('stopwords')
nltk.download('punkt')

# Cleaning the accessibility_caption column
df_merged['cleaned_caption'] = df_merged['accessibility_caption'].replace(
    "No photo description available.", ""
)  # Replace with blank

# Remove "Image may contain:" prefix
df_merged['cleaned_caption'] = df_merged['cleaned_caption'].str.replace(
    r"^Image may contain:\s*", "", regex=True
)  # Remove prefix

# Replace "text that says '...'" with "text"
df_merged['cleaned_caption'] = df_merged['cleaned_caption'].str.replace(r"text that says '.*?'", "text", regex=True)


# Custom stopwords list
custom_stopwords = set([
    "a", "an", "and", "the", "in", "on", "of", "is", "for", "to", "with",
    "by", "this", "that", "at", "as", "one", "more", "many", "or",",","?",","
])

# Function to clean captions
def clean_caption(caption):
    # Ensure the caption is a string
    if not isinstance(caption, str):
        return ''  # Return empty string if it's not a valid string

    # Remove numbers
    caption = re.sub(r'\b\d+\b', '', caption)

    # Tokenize and remove stopwords
    words = caption.split()
    filtered_words = [word for word in words if word.lower() not in custom_stopwords]

    # Rejoin the filtered words into a cleaned string
    return ' '.join(filtered_words)


# Apply the cleaning function to the accessibility_caption column
df_merged['cleaned_caption'] = df_merged['cleaned_caption'].apply(clean_caption)

# Load SpaCy model
nlp = spacy.load("en_core_web_sm")

# Function to lemmatize text
def lemmatize_caption_spacy(caption):
    doc = nlp(caption)
    return ' '.join([token.lemma_ for token in doc if token.is_alpha and not token.is_stop])

# Apply the function to the column
df_merged['cleaned_caption'] = df_merged['cleaned_caption'].apply(lemmatize_caption_spacy)


[nltk_data] Downloading package stopwords to /root/nltk_data...
[nltk_data]   Package stopwords is already up-to-date!
[nltk_data] Downloading package punkt to /root/nltk_data...
[nltk_data]   Package punkt is already up-to-date!


Identifies the 10 most frequent words from the captions and create new one-hot encoded columns for each word to represent their presence in individual captions.

First, it extracts the top 10 frequent words from the processed cleaned_caption text. Then, it defines a function to check if a specific word exists in a given caption. For each of the top 10 words, a new column is added to the DataFrame, where the value is 1 if the word is present in the caption and 0 otherwise.

This transforms the textual data into a structured format that can be easily used for further analysis or modeling.

In [ ]:
from collections import Counter

# Combine all captions into one string
all_captions = ' '.join(df_merged['cleaned_caption'])

# Tokenize into words
all_words = all_captions.split()

# Count the frequency of each word
word_counts = Counter(all_words)

# Get the 10 most frequent unique words along with their frequencies
most_frequent_unique_words = word_counts.most_common(10)

# Print the results
print("Top 10 most frequent unique words with frequencies:")
for word, freq in most_frequent_unique_words:
    print(f"{word}: {freq}")


Top 10 most frequent unique words with frequencies:
person: 6363
people: 5701
outdoor: 4120
stand: 1957
indoor: 1636
sky: 1489
nature: 1256
shoe: 1194
closeup: 1116
food: 1087


In [ ]:
# List of the top 10 most frequent words
top_10_words = [word for word, count in most_frequent_unique_words]

# Function to perform one-hot encoding for each word
def one_hot_encode(caption, word):
    return 1 if word in caption.split() else 0

# Create one-hot encoded columns for each top word
for word in top_10_words:
    df_merged[f"contains_{word}"] = df_merged['cleaned_caption'].apply(lambda x: one_hot_encode(x, word))

In [ ]:
df_merged.head()

username              post_id  \
0  susannahkellogg  1992977640461680979   
1       la.brandon  1673289887866232073   
2        kailovera  1071828292149529960   
3      cowboytails  1735717259248626469   
4            noore  1732054186861771249   

                                             caption  likes  comments  \
0  Some days I just live for naptime!  How many o...    165         0   
1  Winter essential. I use this every other day t...    345         7   
2  Saliendo... Vaya fecha para tomar un vuelo... ...     59         0   
3  From last night’s birthday festivities @luckyd...    112         5   
4         Hey I’m at the store do you need anything?   2194       299   

   comments_parent   timestamp  sponsorship  usertags location  ...  \
0                0  1551801470            0         1       US  ...   
1                7  1513691720            0         1      NaN  ...   
2                0  1441991910            0         0       VE  ...   
3                5  1521133643            0         0       US  ...   
4              299  1520696970            0         1       US  ...   

  contains_person  contains_people contains_outdoor  contains_stand  \
0               0                1                0               0   
1               0                0                0               0   
2               0                0                0               0   
3               0                0                0               0   
4               1                0                0               0   

   contains_indoor  contains_sky contains_nature  contains_shoe  \
0                0             0               0              0   
1                0             0               0              0   
2                0             0               0              0   
3                1             0               0              0   
4                0             0               0              1   

   contains_closeup  contains_food  
0                 0              0  
1                 0              0  
2                 0              0  
3                 0              0  
4                 0              0  

[5 rows x 53 columns]

In [ ]:
output_csv_path = "training_data_and_features.csv"

Write the new Dataset with the new features into CSV file

In [ ]:
df_merged.to_csv(output_csv_path, index=False)

print(f"Data successfully written to {output_csv_path}")

Data successfully written to training_data_and_features.csv


# **SPLIT DATASET AND TRAIN MODEL**

In [ ]:
training_data_csv_path = "training_data_and_features.csv"

In [ ]:
df_training_data = pd.read_csv(training_data_csv_path)

In [ ]:
len(df_training_data)

19997

In [ ]:
print(df_training_data.columns.tolist())

['username', 'post_id', 'caption', 'likes', 'comments', 'comments_parent', 'timestamp', 'sponsorship', 'usertags', 'location', 'accessibility_caption', 'is_verified', 'Username', '#Followers', '#Followees', '#Posts', 'decoded_timestamp', 'Category_beauty', 'Category_family', 'Category_fashion', 'Category_fitness', 'Category_food', 'Category_interior', 'Category_other', 'Category_pet', 'Category_travel', 'time_of_day_Afternoon', 'time_of_day_Evening', 'time_of_day_Morning', 'time_of_day_Night', 'season_Autumn', 'season_Spring', 'season_Summer', 'season_Winter', 'location_likes_mean', 'caption_length', 'word_count', 'hashtag_count', 'mention_count', 'caption_cleaned', 'sentiment_polarity', 'sentiment_subjectivity', 'cleaned_caption', 'contains_person', 'contains_people', 'contains_outdoor', 'contains_stand', 'contains_indoor', 'contains_sky', 'contains_nature', 'contains_shoe', 'contains_closeup', 'contains_food']


In [ ]:
columns_to_use = [
    'comments', 'sponsorship', 'usertags', 'is_verified', '#Followers', '#Followees', '#Posts',
    'Category_beauty', 'Category_family', 'Category_fashion', 'Category_fitness', 'Category_food',
    'Category_interior', 'Category_other', 'Category_pet', 'Category_travel', 'time_of_day_Afternoon',
    'time_of_day_Evening', 'time_of_day_Morning', 'time_of_day_Night', 'season_Autumn', 'season_Spring',
    'season_Summer', 'season_Winter', 'location', 'caption_length', 'word_count', 'hashtag_count',
    'mention_count', 'sentiment_polarity', 'sentiment_subjectivity', 'contains_person',
    'contains_people', 'contains_outdoor', 'contains_stand', 'contains_indoor', 'contains_sky', 'contains_nature',
    'contains_shoe', 'contains_closeup', 'contains_food'
]

In [ ]:
len(columns_to_use)

41

In [ ]:
from sklearn.model_selection import train_test_split

# Define features (X) and target (y)
X = df_merged[columns_to_use]  # Drop the target column from features
y = df_merged['likes']                # Target column

# Step 1: Split into training and temporary sets (validation + testing)
X_train, X_temp, y_train, y_temp = train_test_split(X, y, test_size=0.3, random_state=42)

# Step 2: Split the temporary set into validation and testing
X_val, X_test, y_val, y_test = train_test_split(X_temp, y_temp, test_size=0.5, random_state=42)

In [ ]:
print(f"Training set size: {X_train.shape[0]}")
print(f"Validation set size: {X_val.shape[0]}")
print(f"Testing set size: {X_test.shape[0]}")

Training set size: 13997
Validation set size: 3000
Testing set size: 3000


### **AVOID DATA LEAKAGE**

The `location_likes_mean` feature is calculated exclusively using the training set to avoid data leakage. This ensures that the test and validation sets do not gain insights from the training data. Once computed, this feature is mapped from the training set to the test and validation sets, preserving the integrity of the evaluation process.

In [ ]:
X_y_train = pd.concat([X_train['location'], y_train], axis=1)

In [ ]:
location_likes_mean = X_y_train.groupby('location')['likes'].mean().to_dict()

In [ ]:
type(location_likes_mean)

dict

In [ ]:
location_likes_mean_df = pd.DataFrame(list(location_likes_mean.items()), columns=["location", "location_likes_mean"])

In [ ]:
X_train = pd.merge(
    X_train,
    location_likes_mean_df,
    left_on='location',
    right_on='location',
    how='left'
)

In [ ]:
X_test = pd.merge(
    X_test,
    location_likes_mean_df,
    left_on='location',
    right_on='location',
    how='left'
)

In [ ]:
X_val = pd.merge(
    X_val,
    location_likes_mean_df,
    left_on='location',
    right_on='location',
    how='left'
)

In [ ]:
X_train = X_train.drop(columns=['location'])
X_test = X_test.drop(columns=['location'])
X_val = X_val.drop(columns=['location'])

We fill missing values with the mean of each column to make sure the data is complete and can be used by the model. This helps maintain the overall balance of the data and ensures the model works consistently without being affected by null values.

In [ ]:
from sklearn.impute import SimpleImputer

# Impute missing values with the mean of each column
imputer = SimpleImputer(strategy='mean')
X_train_imputed = imputer.fit_transform(X_train)
X_val_imputed = imputer.transform(X_val)
X_test_imputed = imputer.transform(X_test)

In [ ]:
print(f"Training set size: {X_train_imputed.shape[0]}")
print(f"Validation set size: {X_val_imputed.shape[0]}")
print(f"Testing set size: {X_test_imputed.shape[0]}")

Training set size: 13997
Validation set size: 3000
Testing set size: 3000


**TRAIN WITH RANDOM PARAMETERS**

In [ ]:
from sklearn.ensemble import RandomForestRegressor

# Initialize the model
model = RandomForestRegressor(n_estimators=100, random_state=42)

# Train the model
model.fit(X_train_imputed, y_train)


RandomForestRegressor(random_state=42)

In [ ]:
from sklearn.metrics import mean_squared_error, mean_absolute_error, r2_score

# Predictions on the validation set
y_val_pred = model.predict(X_val_imputed)

# Validation metrics
mse_val = mean_squared_error(y_val, y_val_pred)
mae_val = mean_absolute_error(y_val, y_val_pred)
r2_val = r2_score(y_val, y_val_pred)

print(f"Validation MSE: {mse_val:.2f}")
print(f"Validation MAE: {mae_val:.2f}")
print(f"Validation R^2: {r2_val:.2f}")

Validation MSE: 145469373.82
Validation MAE: 2139.07
Validation R^2: 0.66


# **RESULTS:**

These metrics suggest that the model captures approximately 66% of the variance in the number of likes, which is a reasonable fit for this type of data. However, the relatively high MAE indicates that the model's predictions deviate from the actual values by an average of about 2,139 likes, suggesting room for improvement in fine-tuning the model or refining the features.

### **FEATURE IMPORTANCE**

In this step, we identify the features that have the biggest impact on our model's predictions. By analyzing the importance scores, we can see which factors play the most significant role in determining the outcome. These key features will be the ones we include in our model to ensure it focuses on the most relevant information.

In [ ]:
importances = model.feature_importances_

# Create a DataFrame with feature names and their corresponding importance scores
feature_importance_df = pd.DataFrame({
    'feature': X_train.columns,
    'importance': importances
})

# Sort the DataFrame by importance in descending order
feature_importance_df = feature_importance_df.sort_values(by='importance', ascending=False)

# Display the top 10 most important features
print(feature_importance_df.head(20))

                   feature  importance
4               #Followers    0.595634
0                 comments    0.259247
24          caption_length    0.036960
6                   #Posts    0.018474
5               #Followees    0.015594
25              word_count    0.011563
28      sentiment_polarity    0.008028
17     time_of_day_Evening    0.006701
27           mention_count    0.005276
40     location_likes_mean    0.004069
26           hashtag_count    0.004048
29  sentiment_subjectivity    0.003495
2                 usertags    0.003272
9         Category_fashion    0.003193
36         contains_nature    0.002689
20           season_Autumn    0.002242
31         contains_people    0.002177
13          Category_other    0.001594
35            contains_sky    0.001589
23           season_Winter    0.001471


In [ ]:
new_columns_to_use = feature_importance_df.head(20)['feature'].tolist()

In [ ]:
len(new_columns_to_use)

20

In [ ]:
selected_indices = [X_train.columns.tolist().index(col) for col in new_columns_to_use]

In [ ]:
new_X_train = X_train_imputed[:, selected_indices]

In [ ]:
new_X_val_imputed = X_val_imputed[:, selected_indices]

In [ ]:
selected_column_names = [X_train.columns.tolist()[i] for i in selected_indices]

# Print the column names
print("Selected column names:", selected_column_names)

Selected column names: ['#Followers', 'comments', 'caption_length', '#Posts', '#Followees', 'word_count', 'sentiment_polarity', 'time_of_day_Evening', 'mention_count', 'location_likes_mean', 'hashtag_count', 'sentiment_subjectivity', 'usertags', 'Category_fashion', 'contains_nature', 'season_Autumn', 'contains_people', 'Category_other', 'contains_sky', 'season_Winter']


## **HYPERPARAMETERS**

This process identifies the optimal hyperparameters for the Random Forest model, improving its performance on unseen data.

In [ ]:
from sklearn.model_selection import GridSearchCV

# Initialize the model
rf_model = RandomForestRegressor(random_state=42)

# Define the hyperparameter grid
param_grid = {
    'n_estimators': [100, 200, 300],
    'max_depth': [None, 10, 20],
    'min_samples_split': [2, 5, 10]
}

# Perform grid search
grid_search = GridSearchCV(estimator=rf_model, param_grid=param_grid, scoring='neg_mean_squared_error', cv=3)
grid_search.fit(new_X_train, y_train)

# Best parameters from grid search
print(f"Best parameters: {grid_search.best_params_}")

Best parameters: {'max_depth': 20, 'min_samples_split': 2, 'n_estimators': 300}


In [ ]:
y_val_pred = grid_search.best_estimator_.predict(new_X_val_imputed)

In [ ]:
mse_val = mean_squared_error(y_val, y_val_pred)
mae_val = mean_absolute_error(y_val, y_val_pred)
r2_val = r2_score(y_val, y_val_pred)

print(f"Validation MSE: {mse_val:.2f}")
print(f"Validation MAE: {mae_val:.2f}")
print(f"Validation R^2: {r2_val:.2f}")

Validation MSE: 148829581.17
Validation MAE: 2153.33
Validation R^2: 0.66


In [ ]:
new_X_test_imputed = X_test_imputed[:, selected_indices]

In [ ]:
# Use the best model from grid search
final_model = grid_search.best_estimator_

# Predict on the testing set
y_test_pred = final_model.predict(new_X_test_imputed)

In [ ]:
mse_test = mean_squared_error(y_test, y_test_pred)
mae_test = mean_absolute_error(y_test, y_test_pred)
r2_test = r2_score(y_test, y_test_pred)

print(f"Testing MSE: {mse_test:.2f}")
print(f"Testing MAE: {mae_test:.2f}")
print(f"Testing R^2: {r2_test:.2f}")

Testing MSE: 472847342.79
Testing MAE: 2456.39
Testing R^2: 0.58
